grug/app/src/ctx.rs

```diff
@@ -0,0 +1,110 @@
+use {
+    crate::{Buffer, GasTracker, Shared, Vm},
+    grug_types::{BlockInfo, Storage, Undefined},
+};
+
+/// Wrapper `ctx` struct that holds the `VM`, `storage`, `block`, and `gas_tracker`
+#[derive(Clone)]
+pub struct AppCtx<VM = Undefined<()>, S = Box<dyn Storage>> {
+    pub block: BlockInfo,
+    pub chain_id: String,
+    pub gas_tracker: GasTracker,
+    pub storage: S,
+    vm: VM,
+}
+
+impl AppCtx {
+    pub fn new<VM, S, C>(
+        block: BlockInfo,
+        chain_id: C,
+        gas_tracker: GasTracker,
+        storage: S,
+        vm: VM,
+    ) -> AppCtx<VM, S>
+    where
+        C: Into<String>,
+    {
+        AppCtx {
+            block,
+            chain_id: chain_id.into(),
+            gas_tracker,
+            storage,
+            vm,
+        }
+    }
+}
+
+impl<VM> AppCtx<VM>
+where
+    VM: Vm,
+{
+    pub fn vm(&self) -> &VM {
+        &self.vm
+    }
+
+    /// Downcast the `AppCtx<VM>` to `AppCtx<Undefined<()>>`
+    pub fn downcast(&self) -> AppCtx<Undefined<()>> {
```

Comment:
No need to clone here. The method can take a `self` and return an `AppCtx<Undefined>`.